### ISSUE TITLE
With -n avoid injecting blank lines (and unnecessary newlines)

### ISSUE THREAD

**Issue by DamonHD** [giakki/uncss#331]
It seems a shame when trimming the CSS to a minimum to leave in (in fact, add!) redundant blank lines and newlines; all the filtered CSS should be on a single line with no breaks.

(Thanks for an interesting tool: I'm writing up my comparison for my site against purifycss here: http://www.earth.org.uk/note-on-site-technicals-4.html#uncss )

Rgds

Damon

**Comment by RyanZim** [comment]
@DamonHD I ran some quick tests and I'm not seeing the behavior you described. Can you post a minimal reproducible test case?

**Comment by DamonHD** [comment]
% cat .note-on-carbon-cost-of-CDN.html | uncss -t 0 -n -m screen,print -i .container,.sidebar,.sml,.noprint,.ad -s img/css/base-20171001.css,img/css/desktop-20170906.css,img/css/fullw-20170606.css,img/css/table-20170622.css
    h1,h2,h3{font-family:sans-serif}.sidebar{color:#000;background-color:#cfc;margin:1em 0}.sml{font-size:small}.pgdescription{font-weight:700;margin:1em 0}.ad{border:1px dotted red;margin:1em;padding:3px}.resp{max-width:100%;height:auto}.respfloatr{float:right;clear:right;height:auto;margin:1em 0 1em 1.5em}.respfloatr{max-width:50%}.respfloatr{border-radius:4px}@media screen and (min-width:640px){.container{max-width:800px;margin:auto}}@media print{.ad,.noprint{display:none!important}} 
    @media screen and (min-width:800px){.container{font-size:1.2rem}}
 
    .fullwcdisp{width:100%;width:100vw;position:relative;left:50%;right:50%;margin-left:-50%;margin-left:-50vw;margin-right:-50%;margin-right:-50vw}.fullwcdisp{clear:both;text-align:center}
 
    table.tb1,table.tb1 td,table.tb1 th{border:1px solid}

1) Note the blank lines between preserved CCS fragments for the separate files (currently I am stripping this out with an additional step in the pipeline.
2) Secondary issue: note the trailing \n *not* being stripped from the end of each file's preserved content even when it could be.

**Comment by RyanZim** [comment]
If each of your files contain a trailing newline, uncss will preserve that.

**Comment by DamonHD** [comment]
1) Uncss is adding a newline of its own between files (esp if all CSS dropped from that file) it seems.
2) It's a shame for a minifier not to apply its built-in rules to eliminate trailing whitespace!  Another tool that shall remain nameless does eliminate whitespace between files (where safe).

Rgds

Damon

**Comment by RyanZim** [comment]
I'm still unable to reproduce; I'll need you to make a reduced test case repo for me to be able to investigate.

**Comment by DamonHD** [comment]
If I bundled up my input files and the result in a ZIP, along with the version output, would that be helpful?

FWIW I am getting the same behaviour on a RPi (ie Linux) and on macOS, so I don't *think* that it's a weird one-off!

Rgds

Damon

**Comment by RyanZim** [comment]
I'd prefer if you pushed the files to a github repository.

**Comment by DamonHD** [comment]
OK.  A new one for this purpose?

**Comment by RyanZim** [comment]
Yep

**Comment by DamonHD** [comment]
OK, see if this helps:

https://github.com/DamonHD/201710-uncss-bug

Rgds

Damon

**Comment by RyanZim** [comment]
OK, now I see what you mean. I assumed you were saying uncss inserts blank lines between stylesheets; it doesn't do that, it just puts each sheet on a new line. I'd be open to changing this to not add the newline when `-n` is passed, though I don't think uncss should ever trim existing trailing whitespace from stylesheets. 

This would be a breaking change, but we're currently working on a semver-major release anyhow, so I'll try to get it in.

**Comment by RyanZim** [comment]
For my own reference, we need to change https://github.com/giakki/uncss/blob/master/src/uncss.js#L143 to fix this.

**Comment by DamonHD** [comment]
Why don't you think trimming excess whitespace (leading, trailing, etc) in a non-breaking way isn't the obvious thing to do when you are otherwise helping removing redundancy and doing some fairly deep analysis?  I'm puzzled!  B^>

Anyhow, thanks for looking into this!

Rgds

Damon

**Comment by mikelambert** [comment]
Some people want to use this on their existing stylesheets, to remove old/deprecated code they no longer need.

But also, I'm a fan of "each tool does one thing well". There are other tools that trim whitespace, minimize, or do other space-saving operations. I'd rather them do that (with the plethora of configuration options they then require), rather than have this slowly grow ("oh hey, I noticed you trim newlines at the end, can you trim newlines in the middle of stylesheets too, please? You're already trimming, so I don't see why you wouldn't want to trim more useless stuff. kthxbye").

I was a fan of [unix's small tools](http://www.tldp.org/LDP/GNU-Linux-Tools-Summary/html/c1089.htm) that then get combined with pipes, and view PostCSS and this JS ecosystem somewhat similarly.

**Comment by DamonHD** [comment]
Yes, that I understand.  And indeed if you look at my example you see there's now three adjacent blocks that could easily be merged, maybe we can talk about that one another time!  B^>

But you're (a) adding new whitespace here that I then need to run an extra process to remove incompletely and (b) as I understand it you're reconstituting the CSS from the semantics, so I'd suggest that *preserving* or re-injecting any non-functional content is the odd thing to do.

Rgds

Damon

**Comment by RyanZim** [comment]
We're not reconstituting the CSS; we're removing parts of it that aren't used by the HTML.

Not inserting a blank newline; yes, I can see doing that. Removing newlines: No. That's what CSS minifiers are for; run a minifier after uncss.

**Comment by DamonHD** [comment]
OK, I hear you.

I'll be happy with uncss not adding a \n and I'll ensure that my new minification process will not leave a trailing one in the fragment .css files.  All should be happy.

Thank you.

Rgds

Damon

**Comment by RyanZim** [comment]
I guess I'll call this PR welcome, since I don't have the time to do this myself.
